synapse/handlers/pagination.py

```diff
@@ -448,6 +448,12 @@ async def get_messages(
 
         if pagin_config.from_token:
             from_token = pagin_config.from_token
+        elif pagin_config.direction == "f":
+            from_token = (
+                await self.hs.get_event_sources().get_start_token_for_pagination(
+                    room_id
+                )
+            )
```

Comment:
I wonder if this should clamp to the user's join or just let them get filtered out?

Reply:
Although I guess that depends on history visibility, so likely doesn't make sense to add that complexity here.

Reply:
Fairly sure the visibility should already be getting filtered later, otherwise you could already read messages you're not supposed to by crafting the start token yourself.

Reply:
@reivilibre I read that right as "no other changes needed"?